File: src/services/threatGraphService.ts
import type { FileAnalysisDetail, ThreatGraphEdge, ThreatGraphNode } from '../types'

export function generateThreatGraph(detail: FileAnalysisDetail): {
  nodes: ThreatGraphNode[]
  edges: ThreatGraphEdge[]
} {
  const nodes: ThreatGraphNode[] = []
  const edges: ThreatGraphEdge[] = []

  const malwareId = detail.scanId
  nodes.push({
    id: malwareId,
    label: detail.malwareFamily,
    type: 'malware',
    color: '#ff1744',
  })

  nodes.push({
    id: `hash_${malwareId}`,
    label: `SHA256:${detail.scanId.slice(0, 12)}…`,
    type: 'hash',
    color: '#9e9e9e',
  })
  edges.push({
    id: `${malwareId}-hash`,
    source: malwareId,
    target: `hash_${malwareId}`,
    label: 'has hash',
  })

  for (const url of detail.contactedUrls) {
    const id = `url_${url}`
    nodes.push({ id, label: url, type: 'url', color: '#00bcd4' })
    edges.push({
      id: `${malwareId}-${id}`,
      source: malwareId,
      target: id,
      label: 'contacts',
      animated: true,
    })
  }

  for (const ip of detail.contactedIps) {
    const id = `ip_${ip}`
    nodes.push({ id, label: ip, type: 'ip', color: '#ff9100' })
    edges.push({ id: `${malwareId}-${id}`, source: malwareId, target: id, label: 'connects to' })
  }

  for (const [tactic, items] of Object.entries(detail.behaviors)) {
    for (const item of items) {
      const id = `behavior_${tactic}_${item.slice(0, 12)}`
      nodes.push({ id, label: item, type: 'behavior', color: '#ff69b4' })
      edges.push({ id: `${malwareId}-${id}`, source: malwareId, target: id, label: 'performs' })
    }
  }

  if (detail.threatGroup) {
    const gid = `group_${detail.threatGroup}`
    nodes.push({
      id: gid,
      label: detail.threatGroup,
      type: 'threat_group',
      color: '#9c27b0',
    })
    edges.push({
      id: `${gid}-${malwareId}`,
      source: gid,
      target: malwareId,
      label: 'attributed to',
    })
  }

  return { nodes, edges }
}

File: backend/services/ThreatGraphService.py
from __future__ import annotations

class ThreatGraphService:
    async def generate_threat_graph(self, scan_result: dict) -> dict:
        nodes = []
        edges = []

        malware_node = {
            'id': scan_result.get('file_hash', 'unknown_hash'),
            'label': scan_result.get('malware_family', 'Unknown Malware'),
            'type': 'malware',
            'color': '#ff1744',
            'data': {
                'hash': scan_result.get('file_hash', ''),
                'family': scan_result.get('malware_family', ''),
                'risk_score': scan_result.get('risk_score', 0),
            },
        }
        nodes.append(malware_node)

        for url in scan_result.get('contacted_urls', []):
            url_node = {'id': f"url_{url}", 'label': url, 'type': 'url', 'color': '#00bcd4'}
            nodes.append(url_node)
            edges.append({'id': f"{malware_node['id']}-{url_node['id']}", 'source': malware_node['id'], 'target': url_node['id'], 'label': 'contacts', 'animated': True})

        for ip in scan_result.get('contacted_ips', []):
            ip_node = {'id': f"ip_{ip}", 'label': ip, 'type': 'ip', 'color': '#ff9100'}
            nodes.append(ip_node)
            edges.append({'id': f"{malware_node['id']}-{ip_node['id']}", 'source': malware_node['id'], 'target': ip_node['id'], 'label': 'connects to'})

        for behavior in scan_result.get('detected_behaviors', []):
            behavior_node = {'id': f"behavior_{behavior}", 'label': behavior, 'type': 'behavior', 'color': '#ff69b4'}
            nodes.append(behavior_node)
            edges.append({'id': f"{malware_node['id']}-{behavior_node['id']}", 'source': malware_node['id'], 'target': behavior_node['id'], 'label': 'performs'})

        threat_group = scan_result.get('threat_group')
        if threat_group:
            group_node = {'id': f"group_{threat_group}", 'label': threat_group, 'type': 'threat_group', 'color': '#9c27b0'}
            nodes.append(group_node)
            edges.append({'id': f"group_{threat_group}-{malware_node['id']}", 'source': group_node['id'], 'target': malware_node['id'], 'label': 'attributed to', 'dashed': True})

        return {'nodes': nodes, 'edges': edges}
